backend/src/stwo/circuit_builder.rs

```diff
@@ -191,3 +259,61 @@ where
         }
     }
 }
+
+pub fn constant_with_next_to_witness_col<T: FieldElement>(
+    expr: &AlgebraicExpression<T>,
+    constant_with_next_list: &mut Vec<usize>,
+) {
+    use AlgebraicBinaryOperator::*;
+    match expr {
+        AlgebraicExpression::Reference(r) => {
+            let poly_id = r.poly_id;
+
+            match poly_id.ptype {
+                PolynomialType::Committed => {}
+                PolynomialType::Constant => match r.next {
+                    false => {}
+                    true => {
+                        constant_with_next_list.push(r.poly_id.id as usize);
+                    }
+                },
+                PolynomialType::Intermediate => {}
+            }
+        }
+        AlgebraicExpression::PublicReference(..) => {}
+        AlgebraicExpression::Number(_) => {}
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
+            left,
+            op: Pow,
+            right,
+        }) => match **right {
+            AlgebraicExpression::Number(_) => {
+                constant_with_next_to_witness_col::<T>(left, constant_with_next_list);
+            }
+            _ => unimplemented!("pow with non-constant exponent"),
+        },
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op: _, right }) => {
+            constant_with_next_to_witness_col::<T>(left, constant_with_next_list);
+            constant_with_next_to_witness_col::<T>(right, constant_with_next_list);
+        }
+        AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op: _, expr }) => {
+            constant_with_next_to_witness_col::<T>(expr, constant_with_next_list);
+        }
+        AlgebraicExpression::Challenge(_challenge) => {}
+    }
+}
+
+// This function creates a list of indices of the constant polynomials that have next references constraint
+pub fn get_constant_with_next_list<T: FieldElement>(analyzed: &Analyzed<T>) -> Vec<usize> {
+    let mut all_constant_with_next: Vec<usize> = Vec::new();
+    for id in analyzed.identities_with_inlined_intermediate_polynomials() {
+        if let Identity::Polynomial(identity) = id {
+            let mut constant_with_next: Vec<usize> = Vec::new();
+            constant_with_next_to_witness_col::<T>(&identity.expression, &mut constant_with_next);
+            all_constant_with_next.extend(constant_with_next)
+        }
+    }
+    all_constant_with_next.sort_unstable();
+    all_constant_with_next.dedup();
+    all_constant_with_next
+}
```

Comment:
This should use a `HashSet` or `BTreeSet` rather than a vector.

Reply:
Also it can be simplified by using `post_visit_expressions_in_identities_mut`.
Then `constant_with_next_to_witness_col` can be removed entirely.

Reply:
cool function! the code is simplified now